clippy_lints/src/field_reassign_with_default.rs

```diff
@@ -0,0 +1,208 @@
+/// for when the pattern type is a tuple.
+fn enumerate_bindings_using_default<'cx, 'hir>(
+    cx: &LateContext<'cx>,
+    block: &Block<'hir>,
+) -> Vec<(usize, Symbol, &'cx TyS<'cx>)> {
+    block
+        .stmts
+        .iter()
+        .enumerate()
+        .filter_map(|(idx, stmt)| {
+            if_chain! {
+                // only take `let ...` statements
+                if let StmtKind::Local(ref local) = stmt.kind;
+                // only take bindings to identifiers
+                if let PatKind::Binding(_, _, ident, _) = local.pat.kind;
+                // only when assigning `... = Default::default()`
+                if let Some(ref expr) = local.init;
+                if let ExprKind::Call(ref fn_expr, _) = &expr.kind;
+                if let ExprKind::Path(qpath) = &fn_expr.kind;
+                if let Res::Def(_, def_id) = qpath_res(cx, qpath, fn_expr.hir_id);
+                // right hand side of assignment is `Default::default`
+                if match_def_path(cx, def_id, &paths::DEFAULT_TRAIT_METHOD);
+                then {
+                    // Get the type of the pattern
+                    let ty = cx.typeck_results().pat_ty(local.pat);
+                    // Ignore tuples
+                    if let ty::Tuple(_) = ty.kind() {
+                        return None;
+                    }
+                    Some((idx, ident.name, ty))
+                }
+                else {
+                    None
+                }
+            }
+        })
+        .collect()
+}
+
+fn stmt_shadows_binding(this: &Stmt<'_>, shadowed: Symbol) -> bool {
+    if let StmtKind::Local(local) = &this.kind {
+        if let PatKind::Binding(_, _, ident, _) = local.pat.kind {
+            return ident.name == shadowed;
+        }
+    }
+    false
+}
+
+/// Returns the reassigned field and the assigning expression (right-hand side of assign).
+fn field_reassigned_by_stmt<'hir>(this: &Stmt<'hir>, binding_name: Symbol) -> Option<(Ident, &'hir Expr<'hir>)> {
+    if_chain! {
+        // only take assignments
+        if let StmtKind::Semi(ref later_expr) = this.kind;
+        if let ExprKind::Assign(ref assign_lhs, ref assign_rhs, _) = later_expr.kind;
+        // only take assignments to fields where the left-hand side field is a field of
+        // the same binding as the previous statement
+        if let ExprKind::Field(ref binding, field_ident) = assign_lhs.kind;
+        if let ExprKind::Path(ref qpath) = binding.kind;
+        if let QPath::Resolved(_, path) = qpath;
+        if let Some(second_binding_name) = path.segments.last();
+        if second_binding_name.ident.name == binding_name;
+        then {
+            Some((field_ident, assign_rhs))
+        } else {
+            None
+        }
+    }
+}
+
+/// Returns the vec of fields for a struct and an empty vec for non-struct ADTs.
+fn fields_of_type<'a>(ty: &'a TyS<'_>) -> Vec<Ident> {
```

Comment:
```suggestion
fn fields_of_type(ty: Ty<'_>) -> Vec<Ident> {
```
This can be simplified